docs/book/v2/migration.md

```diff
@@ -0,0 +1,83 @@
+# Migration from Version 1
+
+This document details changes made from version 1 of this package or from its implementation in
+[laminas-mvc](https://docs.laminas.dev/laminas-mvc/).
+If installed, it overrides the default behavior in laminas-mvc.
+
+## `'controller'` Route Array Key
+
+In addition to providing a `'middleware'` array key in the route defaults definition, the value of the `'controller'`
+array key has to be explicitly set to `Laminas\Mvc\Middleware\PipeSpec::class`:
+```php
+/* ... */
+'some-route' => [
+    'type' => Literal::class,
+    'options' => [
+        'route' => '/some-route',
+        'defaults' => [
+            'controller' => \Laminas\Mvc\Middleware\PipeSpec::class,
+            'middleware' => SomeRequestHandler::class,
+        ],
+    ],
+],
+```
+
+## `PipeSpec` Instance in `middleware` Definition
+
+If you used to pass an array of middlewares to the `middleware` array key to create a middleware pipe, it has to be
+rewritten to an instance of the `PipeSpec` object.
+
+Before:
+
+```php
+/* ... */
+'some-route' => [
+    'type' => Literal::class,
+    'options' => [
+        'route' => '/some-route',
+        'defaults' => [
+            'middleware' => [
+                SomeMiddleware::class,
+                SomeRequestHandler::class,
+            ],
+        ],
+    ],
+],
+```
+
+After:
+
+```php
+/* ... */
+'some-route' => [
+    'type' => Literal::class,
+    'options' => [
+        'route' => '/some-route',
+        'defaults' => [
+            'controller' => \Laminas\Mvc\Middleware\PipeSpec::class,
+            'middleware' => new \Laminas\Mvc\Middleware\PipeSpec(
+                SomeMiddleware::class,
+                SomeRequestHandler::class
+            ),
+        ],
+    ],
+],
+```
+
+## PSR-15 Middleware
+
+This package now supports only PSR-15 interfaces. Support of `http-interop/http-middleware` and `callable` middleware
+has been dropped.
+
+You have to rewrite all your existing handlers and middleware to implement `Psr\Http\Server\RequestHandlerInterface`
+or `Psr\Http\Server\MiddlewareInterface`, respectively.
+
+Please also refer
+to [laminas-stratigility v3 migration guide](https://docs.laminas.dev/laminas-stratigility/v3/migration/)
+if you used any of its features.
+
+## Removed Route Match Params from Request
+
+Previously, for the matched route both the `Laminas\Router\RouteMatch` object as well as individual matched params were
+added to the PSR-7 request as attributes. This is no longer the case and only the `RouteMatch` object can be accessed as
+PSR-7 request attribute: `$request->getAttribute(\Laminas\Router\RouteMatch::class);`
```

Comment:
This should have an example demonstrating _prior usage_, and how users will now access those parameters via the `RouteMatch`.